RFC-0004-jdbc-join-pushdown.md

```diff
@@ -0,0 +1,429 @@
+# **RFC-0004-jdbc-join-pushdown for Presto**
+
+See [CONTRIBUTING.md](CONTRIBUTING.md) for instructions on creating your RFC and the process surrounding it.
+
+## Jdbc join pushdown in presto
+
+Proposers
+
+* Ajas M M
+
+
+## [Related Issues]
+
+https://github.com/prestodb/presto/issues/23152
+
+## Summary
+
+When presto identifies a join query which is specific to a Jdbc remote datasource, it split the join query into multiple select query based on the tables involved in the join query and select all the records from each tables using the jdbc connector without passing the join condition. Then the results of these sub-queries are fetched into the presto workers, additional operations such as filters, joins, sorts are applied before sending back to the user.
+
+If we  "Push down" or send these joins which involves on same catalog/remote datasource as part of the  SQL along with join condition to the remote data source it increase the performance 3x to 10x.
+
+
+
+## Background
+
+
+
+This implementation is to address a performance limitation of Presto federation of SQLs of JDBC connector to remote data sources such as DB2, Postgres, Oracle etc. Currently, in Presto, we have predicate pushdown (WHERE condition pushdown) to some extent in JDBC connectors and not having any join pushdown or join condition pushdown capabilities. 
+
+This cause high performance impact on join queries and it is raised by some of our client. While comparing with competitors we also missing the Jdbc join pushdown capabilities.
+
+We did a poc by changing the presto generated PlanNode to handle jdbc join pushdown and it increases the performance from 3x on postgres and  8x on db2 remote datasource. Now we need to perform its actual implementation 
+
+
+
+## Proposed Implementation
+
+If presto get a join query which is trying to join tables either from same datasource or from different datasource, it is receiving as a string formatted sql query. Using presto parser and analyser, presto validate the syntax and converted to Query (Statement) object. This Query object is converted to presto internal reference architecture called Plan, using its logical and physical optimizers. Finally this plan is executed by the executor. 
+
+Currently for the join query, presto create a final plan which contains separate TableScanNode for each table that participated on join query and this TableScanNode info is used by the connector to create the select query. On top of this select query result, presto apply join condition and other predicate to provide the final result.
```

Comment:
Would be great to add an example to clarify this structure.

Reply:
This structure is explained under `Highlevel Design` as well as on session `3 - JdbcJoinPushdown Rule creation`